deploy/libproject/deploy.rb

```diff
@@ -0,0 +1,207 @@
+    "The Google Cloud project to deploy to."
+  )
+  op.add_option(
+    "--account [account]",
+    lambda {|opts, v| opts.account = v},
+    "Service account to act as for deployment, if any. Defaults to the GAE " +
+    "default service account."
+  )
+  op.add_option(
+    "--key_file [key file]",
+    lambda {|opts, v| opts.key_file = v},
+    "Path to a service account key file to be used for deployment"
+  )
+  op.add_option(
+    "--git_version [git version]",
+    lambda {|opts, v| opts.git_version = v},
+    "GitHub tag or branch, e.g. 'v1-0-rc1', 'origin/master'. Branch names " +
+    "must be prefixed with 'origin/'"
+  )
+  op.add_option(
+    "--app_version [app version]",
+    lambda {|opts, v| opts.app_version = v},
+    "App Engine version to deploy as"
+  )
+  op.add_option(
+    "--promote",
+    lambda {|opts, v| opts.promote = true},
+    "Promote this version to immediately begin serving traffic"
+  )
+  op.add_option(
+    "--no-promote",
+    lambda {|opts, v| opts.promote = false},
+    "Deploy, but do not yet serve traffic from this version - DB migrations are still applied"
+  )
+  op.add_validator lambda {|opts| raise ArgumentError if opts.project.nil?}
+  op.add_validator lambda {|opts| raise ArgumentError if opts.account.nil?}
+  op.add_validator lambda {|opts| raise ArgumentError if opts.promote.nil?}
+
+  op.parse.validate
+
+  common = Common.new
+  unless Workbench::in_docker?
+    source_project = get_source_project(op.opts.project)
+    live_version = get_live_gae_version(op.opts.project)
+    live_source_version = source_project and get_live_gae_version(source_project)
+    if live_source_version and not op.opts.app_version
+      common.status "--app_version defaulting to #{live_source_version}; " +
+                    "found on project #{source_project}"
+      op.opts.app_version = live_source_version
+    end
+    if live_source_version and not op.opts.git_version
+      common.status "--git_version defaulting to #{live_source_version}; " +
+                    "found on project #{source_project}"
+      op.opts.git_version = live_source_version
+    end
+    if not op.opts.git_version or not op.opts.app_version
+      if source_project
+        common.error "Failed to find a promotion candidate version in project "+
+                     source_project
+      else
+        common.error "Project #{op.opts.project} has no environment from which " +
+                     "to promote a release; please verify that you intended to " +
+                     "deploy to this environment using this method (uncommon)"
+      end
+      common.error "No default version found, both --git_version and " +
+                   "--app_version must be specified"
+      exit 1
+    end
+
+    # TODO: Might be nice to emit the last version creation time here as a
+    # sanity check (need to pick which service to do that for...).
+    common.status "Current live version is '#{live_version}' (project " +
```

Comment:
Is there a prompt here to confirm? (It's useful for somebody to verify that we're promoting the right version manually when pushing to stable / prod.)

Reply:
Added a prompt.